Add automatic generation of CRDs for CNP and CCNP

christarazi opened the pull request:
This paves the path for generating CRDs using [controller-tools](https://github.com/kubernetes-sigs/controller-tools)'s [controller-gen tool](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/controller-gen). This will help in preparation for the Kubernetes 1.19 release. A significant benefit is the ability to generate the validation schemas, instead of writing them by hand, which can be error-prone.

See commit msgs.

Updates https://github.com/cilium/cilium/pull/11476
Related https://github.com/cilium/cilium/issues/11142

aanm committed:
k8s, policy: Add automated CRD generation

This commit provides a replacement for handwritten CRD schema validation
via annotating struct fields with markers for use with the
controller-gen tool.

Running the tool and generating the binary representation have been
automated in these Makefile targets:

  1) `make manifests` which will run the automatic CRD generator tool
     and outputs CRD YAMLs into `examples/crds/`.

  2) `make go-bindata` which will embed the binary representation of the
     CRD YAMLs so that it can be imported directly into Cilium.

Note that this change bring back go-bindata, a tool that we previously
removed in #10177. This can be swapped in the future if we decide
there's a better alternative.

Commits following this will add the bindata.go file which contains the
binary representation of the generated CRD.

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Generate CCNP CRD with controller-gen

Fakes and deepcopy generated code have also been updated.

This CRD is generated using
https://book.kubebuilder.io/reference/generating-crd.html.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s, policy: Make docs consistent on CRD types

This commit helps keep the description field in the CRD to have
consistent formatting across all the types. It also moves all multi-line
marker comments together akin to the formatting from the core K8s CRD
files (pkg/k8s/**/k8s/apis/core/v1/types.go).

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Update and generate CiliumNetworkPolicy CRD

For unknown reasons, we need to manually add the
metadata.creationTimestamp column (as "Age" in kubectl output) even
though this is not a Cilium-specific field (it is part of
metav1.ObjectMeta type). Thus, we are omitting adding the description to
this column because we want to avoid hardcoding the documentation of the
field.

This CRD is generated using
https://book.kubebuilder.io/reference/generating-crd.html.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Separate out CiliumNetworkPolicy CRD type

This commit separates out CiliumNetworkPolicy from v2/types.go to
prepare for use with the controller-gen tool for CRD generation.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Separate CiliumClusterwideNetworkPolicy CRD

This commit separates out CiliumClusterwideNetworkPolicy (CCNP) from
v2/types.go to prepare for use with the controller-gen tool for
automatic CRD generation.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
vendor: Add go-bindata

This is used for embedding the generated CRD YAMLs into binary form.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
api, policy: Export fields in CRD / API structs

The reason for this commit is because the controller-gen tool used for
CRD generation requires fields in the a CRD struct to be exported. In
addition, the tool requires that all struct fields have JSON tags.

This commit temporarily exports these fields and leaves indicators for
future refactor.

Once https://go-review.googlesource.com/c/tools/+/245857 is merged, this
commit can be reverted.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
vendor: add crd tool generator

To generate CRD validation schemas (also known as structual validation
schemas[1]), we bring in controller-tools. This houses a tool called
controller-gen which performs various K8s related code generation.

Note that we are using a fork of controller-tools due to the upstream
(as of v0.3.0) lacking support in two main areas.

1) A type which implements custom JSON marshalling has its validation
   schema generation skipped. In our case, api.Rule (which is set as the
   Spec field in CNP and CCNP) implements custom JSON marshalling. With
   the upstream tool, this skips the validation generation for it. In
   our fork, we revert this feature. See PR:
   https://github.com/kubernetes-sigs/controller-tools/pull/427
2) For CNP and CCNP, we require `oneOf` validation for the
   endpointSelector and nodeSelector fields. This is because we want to
   validate that either endpointSelector OR nodeSelector is present, but
   not both, and at least one. In the upstream, there's currently an
   open PR that implements support for this. We have manually merged it
   in our fork. See PR:
   https://github.com/kubernetes-sigs/controller-tools/pull/298

This commit also brings updates to our dependencies:

- gomega v1.7.0 -> v1.8.1
- golang.org/x/internal/{tools,event...}

[1]:
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Use slim metav1.Time as a timestamp

This allows us to avoid any JSON marshaling issues with controller-gen
tool, and gets rid of our handmade Timestamp type.

Note that this commit adds the extra JSON tag, as that will be important
when we start using the controller-gen tool to automate the generation
of CRDs. Without the tag, the tool will complain about struct fields
which don't have JSON tags.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Retrieve CRDs in binary form via helper

This commit introduces a function called GetPregeneratedCRD.

Now that we generate CRDs with the controller-gen tool, it allows us to
also generate validation schema as well. This commit refactors how we
retrieve the pregenerated CRDs. This refactor allows us to pass the
pregenerated CRD to the validator package.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/azure/types/types.go:
```diff
@@ -113,17 +113,17 @@ type AzureInterface struct {
-	// resourceGroup is the resource group the interface belongs to
-	resourceGroup string
+	// ResourceGroup is the resource group the interface belongs to
+	ResourceGroup string `json:"resource-group,omitempty"`
```
they were not exported, so they should be `json:"-"`

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/cnp_types.go:
```diff
@@ -157,12 +157,10 @@ func NewTimestamp() Timestamp {
-}
+type Timestamp metav1.Time
 
 func (in *Timestamp) DeepEqual(other *Timestamp) bool {
```
I think we can replace this `Timestamp` struct, i.e. remove it entirely, and replace all usages with a `github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1.Time` instead.

christarazi replied:
Done

aanm replied:
It's not addressed.

christarazi replied:
This was addressed in a later commit which replaces the Timestamp type. This commit is just a "rename".

aanm commented on operator/k8s_node.go:
```diff
@@ -162,7 +161,7 @@ func runCNPNodeStatusGC(name string, clusterwide bool, ciliumNodeStore *store.Sh
 			RunInterval: operatorOption.Config.CNPNodeStatusGCInterval,
 			DoFunc: func(ctx context.Context) error {
-				lastRun := time.Now().Add(-operatorOption.Config.NodesGCInterval)
+				lastRun := cilium_v2.NewTimestamp().Add(-operatorOption.Config.NodesGCInterval)
```
This could be done in a different commit, like the one where you replaced the timestamp with metav1.Time

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -526,990 +418,22 @@ func needsUpdate(clusterCRD *apiextensionsv1beta1.CustomResourceDefinition) bool
-			},
-		},
-	}
+	// TODO(christarazi): Is CNPCRV still needed? It only exists for the
```
Yes, it is still needed, it's used to perform the validation in the preflight command

aanm reviewed (changes requested):
:100: looks really good! One other thing I noticed that is missing is the lack of the annotation field containing the version number.

christarazi committed:
policy: Validate entities at the CRD level

This commit allows us to reject policies with invalid entities at the
apiserver level as soon as a user has applied the YAML.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Disallow nodeSelector for CNP at agent level

Previously, this validation was done at the apiserver level with
handwritten schema validation. Now that we have automated CRD schema
validation, we must implement this check at the agent level as a
stopgap.

The reason for this is that given the relationship of CCNP and CNP (CCNP
embeds CNP), it was impossible to represent the following in
controller-gen marker comments to generate validation schema:

- CNP only supports EndpointSelector.
- CCNP supports both EndpointSelector and NodeSelector, but only one at
  a time in a policy.

This commit is meant to be reverted once an approach for seperating the
CNP and CCNP CRD types has been implemented.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/azure/api/mock/mock.go:
```diff
@@ -200,7 +200,7 @@ func (a *API) AssignPrivateIpAddressesVMSS(ctx context.Context, vmName, vmssName
 		}
 
-		if intf.Name != interfaceName || intf.VMID() != vmName {
+		if intf.Name != interfaceName || intf.GetVMID() != vmName {
```
Please split this commit in 2. One that only adds the `json:"-"` and another that renames all the fields. The field renaming commit will likely be reverted in the future once https://go-review.googlesource.com/c/tools/+/245857 is merged.

christarazi replied:
Done

aanm commented on pkg/k8s/apis/cilium.io/v2/cnp_types.go:
```diff
@@ -0,0 +1,320 @@
+// This file has been created to sperate the types defined in v2/types.go so
+// that we can control the files processed by controller-gen for generating
+// CRDs.
+
```
You can remove this, there's not need to justify files IMO.

christarazi replied:
Removed

aanm commented on pkg/k8s/apis/cilium.io/v2/ccnp_types.go:
```diff
@@ -0,0 +1,53 @@
+// This file has been created to sperate the types defined in v2/types.go so
+// that we can control the files processed by controller-gen for generating
+// CRDs.
+
```
You can remove this, there's not need to justify files IMO.

christarazi replied:
Removed

aanm commented on pkg/k8s/slim/k8s/apis/meta/v1/time.go:
```diff
@@ -30,7 +30,7 @@ import (
 // +protobuf.as=Timestamp
 // +protobuf.options.(gogoproto.goproto_stringer)=false
 type Time struct {
-	time.Time `protobuf:"-"`
```
I don't think this `json:"-"` is really, is it?

christarazi replied:
`controller-gen` complains about a field missing a JSON tag if I don't have this change.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -248,18 +246,18 @@ type CiliumIdentityList struct {
 
-// AddressPairList is a list of address pairs
+// AddressPairList is a list of address pairs.
 // +k8s:deepcopy-gen=false
```
Inconsistent with other structs, discussed offline.

christarazi replied:
Done

aanm commented on pkg/policy/api/http.go:
```diff
@@ -37,26 +37,28 @@ const (
+	// The secret must only contain one entry. If the referred secret does not
+	// exist, and there is no "Value" specified, the match will fail.
 	//
 	// +optional
```
Similar to what is done with `Value string `json:"value,omitempty"`` on line 62.
```
	//
	// +kubebuilder:validation:Optional
```

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -82,10 +83,15 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
-	)
+	crdBytes, err := examplesCrdsCiliumnetworkpoliciesYamlBytes()
+	if err != nil {
+		panic(err)
```
`return err`

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -82,10 +83,15 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
+	ciliumCRD := apiextensionsv1beta1.CustomResourceDefinition{}
+	err = yaml.Unmarshal(crdBytes, &ciliumCRD)
+	if err != nil {
+		panic(err)
```
`return err`

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/cnp_types.go:
```diff
@@ -34,10 +34,14 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// +genclient
```
This was accidental removed 2 commits behind.

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/ccnp_types.go:
```diff
@@ -22,15 +22,29 @@ import (
+	// https://github.com/kubernetes-sigs/controller-tools/blob/4a903ddb7005459a7baf4777c67244a74c91083d/pkg/crd/gen.go#L221
+
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
```
This needs to be in a separate commit because it's introducing bugs. With master branch, the json unmarshaling is happening into the fields of the embedded CiliumNetworkPolicy structure, with this change that is no longer happening which causes the following code to fail:
```golang
// DeepEqual compares 2 CNPs while ignoring the LastAppliedConfigAnnotation and
// ignoring the Status field of the CNP.
func (in *CiliumNetworkPolicy) DeepEqual(other *CiliumNetworkPolicy) bool {
	switch {
	case (in == nil) != (other == nil):
		return false
	case (in == nil) && (other == nil):
		return true
	}

	if !(in.Name == other.Name && in.Namespace == other.Namespace) {
		return false
	}

	// Ignore v1.LastAppliedConfigAnnotation annotation
	lastAppliedCfgAnnotation1, ok1 := in.GetAnnotations()[v1.LastAppliedConfigAnnotation]
	lastAppliedCfgAnnotation2, ok2 := other.GetAnnotations()[v1.LastAppliedConfigAnnotation]
	defer func() {
		if ok1 && in.GetAnnotations() != nil {
			in.GetAnnotations()[v1.LastAppliedConfigAnnotation] = lastAppliedCfgAnnotation1
		}
		if ok2 && other.GetAnnotations() != nil {
			other.GetAnnotations()[v1.LastAppliedConfigAnnotation] = lastAppliedCfgAnnotation2
		}
	}()
	delete(in.GetAnnotations(), v1.LastAppliedConfigAnnotation)
	delete(other.GetAnnotations(), v1.LastAppliedConfigAnnotation)

	return comparator.MapStringEquals(in.GetAnnotations(), other.GetAnnotations()) &&
		in.deepEqual(other)
}
```

If we call the DeepEqual method from a CCNP structure, it will use the DeepEqual method of the CNP since it's embedded, as expected. The problem rises that the CNP's DeepEqual is accessing fields of itself, so `in.Name`, `in.GetAnnotations()`, etc, will always be empty because the unmarshalling happen into the `metav1.ObjectMeta` and `metav1.TypeMeta` of the CCNP and not the CNP like it was previously.

A suggestion to fix this is to have a DeepEqual wrapper in the CCNP which has similar code in the CNP's DeepEqual:
```
func (in *CiliumClusterwideNetworkPolicy) DeepEqual(other *CiliumClusterwideNetworkPolicy) bool {
```
(Note: It's probably better to refactor the entire code of the CNP's DeepEqual, except the in.deepEqual(other), into a dedicated function so that we cna use it in the CCNP's DeepEqual and CNP's DeepEqual.)

christarazi replied:
Implemented in 5432c71b8cdb022c9951e7d07606115d1e4e0402. Is this what you had in mind? Ideally this should be tested somewhere (JSON marshalling and unmarshalling), but I couldn't find a single spot for this. Where would you suggest adding a test to validate this?

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -118,13 +118,18 @@ func createCNPCRD(clientset apiextensionsclient.Interface) error {
-	)
+	crdBytes, err := examplesCrdsCiliumclusterwidenetworkpoliciesYamlBytes()
+	if err != nil {
+		panic(err)
```
`return err`

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -118,13 +118,18 @@ func createCNPCRD(clientset apiextensionsclient.Interface) error {
+	ciliumCRD := apiextensionsv1beta1.CustomResourceDefinition{}
+	err = yaml.Unmarshal(crdBytes, &ciliumCRD)
+	if err != nil {
+		panic(err)
```
`return err`

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -118,13 +118,18 @@ func createCNPCRD(clientset apiextensionsclient.Interface) error {
+	ciliumCRD := apiextensionsv1beta1.CustomResourceDefinition{}
+	err = yaml.Unmarshal(crdBytes, &ciliumCRD)
+	if err != nil {
+		panic(err)
```
`return err`

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -80,19 +80,45 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
+	case CCNPCRDName:
+		crdBytes, err = examplesCrdsCiliumclusterwidenetworkpoliciesYamlBytes()
+	default:
+		scopedLog.Panic("Pregenerated CRD does not exist")
```
`s/Panic/Fatal` to crash Cilium.

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -80,19 +80,45 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
+
 	if err != nil {
-		panic(err)
+		scopedLog.WithError(err).Panic("Error retrieving pregenerated CRD")
```
`s/Panic/Fatal` to crash Cilium.

christarazi replied:
Fixed

aanm commented on pkg/k8s/apis/cilium.io/v2/client/register.go:
```diff
@@ -80,19 +80,45 @@ func CreateCustomResourceDefinitions(clientset apiextensionsclient.Interface) er
 	err = yaml.Unmarshal(crdBytes, &ciliumCRD)
 	if err != nil {
-		panic(err)
+		scopedLog.WithError(err).Panic("Error unmarshalling pregenerated CRD")
```
`s/Panic/Fatal` to crash Cilium.

christarazi replied:
Fixed

christarazi committed:
azure, policy: Add JSON tags to CRD fields

JSON tags are required for controller-gen, a tool for generating CRDs.
This commit adds JSON tags to all struct fields that are part of the
API, i.e. CRD.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
k8s: Explicitly embed CRD types into CCNP

This commit explicitly embeds metav1.TypeMeta and metav1.ObjectMeta, as these
are needed for the following commits which introduce automated CRD generation
using the controller-gen tool. Due to the (redundant) embedding, changes were
required in pkg/k8s. We need to embed these fields (for now) because
controller-gen will not detect a CRD type unless it has those fields.

This commit also modifies the deepequal code for CNP and CCNP to
deduplicate the functions.

The deepcopy generated code has also been updated.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/policy/api/selector.go:
```diff
@@ -34,16 +34,25 @@ var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "policy-api")
 	// EndpointSelector. It is populated when EndpointSelectors are created
 	// via `NewESFromMatchRequirements`. It is immutable after its creation.
-	cachedLabelSelectorString string
+	CachedLabelSelectorString string `json:"-"`
```
The addition of the JSON tag should have been done in the previous commit.

christarazi replied:
Fixed

aanm reviewed (approved):
Does not require a second review from my side.

I've created this as follow ups https://github.com/cilium/cilium/issues/12737

aanm reviewed (changes requested):
Sorry, I noticed that we aren't updating the `CustomResourceDefinitionSchemaVersion` version as we are changing the schema.

Also, before we merge this we should double check if the generated CRD is different than the current one. I've spotted a missing regex for the `matchLabels` field as well as it seems the properties of the `endpointSelector` seem to be empty?
```
    schema:
      openAPIV3Schema:
        description: CiliumNetworkPolicy is a Kubernetes third-party resource with
          an extended version of NetworkPolicy.
        properties:
          apiVersion:
            description: 'APIVersion defines the versioned schema of this representation
              of an object. Servers should convert recognized schemas to the latest
              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
            type: string
          kind:
            description: 'Kind is a string value representing the REST resource this
              object represents. Servers may infer this from the endpoint the client
              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
            type: string
          metadata:
            type: object
          spec:
            description: Spec is the desired Cilium specific rule specification.
            oneOf:
            - properties:
                endpointSelector: {}
              required:
              - endpointSelector
            - properties:
                nodeSelector: {}
              required:
              - nodeSelector
```

christarazi committed:
k8s: Bump CRD schema version to 1.22.1

Bumping the version following updates allowing automated schema
generation.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
slim/k8s: Define MatchLabels value type

The reason for this commit is to generate validation for MatchLabels. In
upstream controller-tools, in order to generate validation for maps, a
new type (either for key or value) or a type alias must be defined so
that validation markers (`+kubebuilder...`) can be placed on the type.
There is not upstream support which would allow not having to define new
types. If this changes in the future, we should revert this commit, and
apply the new markers appropriately.

This commit relies on support in our fork of controller-tools:
https://github.com/christarazi/controller-tools/commit/1b051103a13760ad557ad7df5b2c022287ce9ff4

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
contrib: Generate CRDs during k8s code generation

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
docs: Update dev requirements with go-bindata

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
.github: Add go-bindata installation step

Now that the "Check k8s generated files" step includes generating CRDs
and running go-bindata to serialize the CRDs into binary form, we need
to install go-bindata.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm reviewed (approved):
:100:
